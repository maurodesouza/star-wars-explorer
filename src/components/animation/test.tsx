import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Animation, AnimationProps } from '.';

jest.mock('react-lottie', () => ({
  __esModule: true,
  default: function Mock() {
    return <svg data-testid="lottie" />;
  },
}));

const dataMock: AnimationProps = {
  animation: 'loading',
  label: 'foo',
};

describe('<Animation />', () => {
  it('should render the animation styles correct', () => {
    const { container } = renderWithProviders(<Animation {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the animation component correct', () => {
    renderWithProviders(<Animation {...dataMock} />);

    const label = screen.getByText(dataMock.label!);
    const lottie = screen.getByTestId('lottie');

    expect(label).toBeInTheDocument();
    expect(lottie).toBeInTheDocument();
  });
});
