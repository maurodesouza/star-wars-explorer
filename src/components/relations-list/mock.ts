import { Entities } from 'types';
import { getImageUrl } from 'utils';

const mock = {
  relations: [
    {
      id: '1',
      title: 'C-3PO',
      image: getImageUrl(Entities.CHARACTERS, '2'),
      entity: Entities.CHARACTERS,
    },
    {
      id: '1',
      title: 'Star wars',
      image: getImageUrl(Entities.FILMS, '2'),
      entity: Entities.FILMS,
    },
  ],
};

export { mock };
