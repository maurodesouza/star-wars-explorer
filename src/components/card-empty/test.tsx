import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { EmptyCard } from '.';

const dataMock = {
  id: 'test',
};

describe('<EmptyCard />', () => {
  it('should render the empty card styles correct', () => {
    const { container } = renderWithProviders(<EmptyCard {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the empty card component correct', () => {
    renderWithProviders(<EmptyCard {...dataMock} />);

    const label = screen.getByText(`${dataMock.id}#`);

    expect(label).toBeInTheDocument();
  });
});
