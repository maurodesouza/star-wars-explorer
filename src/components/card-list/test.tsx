import { cleanup } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { createArray } from 'utils';

import { CardList } from '.';

jest.mock('components', () => ({
  __esModule: true,
  Card: function Mock() {
    return <div data-testid="card" />;
  },
}));

const itemMock = {
  id: '1',
  title: 'C-3PO',
  image: 'image',
  entity: Entities.CHARACTERS,
};

describe('<CardList />', () => {
  describe('styles', () => {
    beforeAll(() => {
      jest.resetAllMocks();
    });

    it('should render the card list styles correct', () => {
      const { container } = renderWithProviders(<CardList />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('should render the items card correct', () => {
    const counts = [4, 8];

    counts.forEach(count => {
      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const render = renderWithProviders(<CardList items={items} />);
      const findedItems = render.getAllByTestId(/card/i);

      expect(findedItems.length).toBe(count);

      cleanup();
    });
  });

  it('should render the fills correct', () => {
    const minCountToFill = 4;
    const counts = [1, 3, 4, 6];

    counts.forEach(count => {
      const result = count - minCountToFill;
      const lackForFill = result <= 0 ? Math.abs(result) : 0;

      const items = createArray(count).map((_, index) => ({
        ...itemMock,
        id: String(index),
      }));

      const { container } = renderWithProviders(<CardList items={items} />);
      const findedItems = container.firstElementChild!.querySelectorAll(
        'div:not([data-testId]'
      );

      expect(findedItems.length).toBe(lackForFill);
      cleanup();
    });
  });
});
