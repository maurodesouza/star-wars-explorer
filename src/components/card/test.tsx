/* eslint-disable @typescript-eslint/no-var-requires */

import { act, fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { Entities } from 'types';
import { Card } from '.';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');
const push = jest.fn();

useRouter.mockImplementation(() => ({
  push,
}));

const mockData = {
  id: '1',
  label: 'C-3PO',
  image:
    'https://raw.githubusercontent.com/tbone849/star-wars-guide/master/build/assets/img/characters/2.jpg',
};

describe('<Card />', () => {
  it('should render the card correct', () => {
    renderWithProviders(<Card {...mockData} />);

    expect(
      screen.getByRole('img', { name: mockData.label })
    ).toBeInTheDocument();

    expect(
      screen.getByRole('heading', { name: mockData.label })
    ).toBeInTheDocument();
  });

  it("should render the no image found element when can't load image url", async () => {
    renderWithProviders(<Card {...mockData} />);

    const image = screen.getByRole('img', { name: mockData.label });

    act(() => {
      fireEvent.error(image);
    });

    const noImageFoundElement = screen.getByText(/no found/i);

    expect(noImageFoundElement).toBeInTheDocument();
  });

  it('should render the card styles correct', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should navigate to entity page', () => {
    const { container } = renderWithProviders(<Card {...mockData} />);

    fireEvent.click(container.firstChild!);

    expect(push).toBeCalledTimes(1);
    expect(push).toBeCalledWith(`/${Entities.CHARACTERS}/${mockData.label}`);
  });
});
