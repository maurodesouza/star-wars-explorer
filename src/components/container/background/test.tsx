import { renderWithProviders } from 'utils/test/helpers';
import { Background } from '.';

describe('components/container', () => {
  describe('<Background />', () => {
    it('should render the background styles correct', () => {
      const { container } = renderWithProviders(<Background />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
