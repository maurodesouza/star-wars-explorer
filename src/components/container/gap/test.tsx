import { renderWithProviders } from 'utils/test/helpers';
import { Gap } from '.';

describe('components/container', () => {
  describe('<Gap />', () => {
    it('should render the gap styles correct', () => {
      const { container } = renderWithProviders(<Gap />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
