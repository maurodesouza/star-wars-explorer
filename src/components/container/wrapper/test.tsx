import { renderWithProviders } from 'utils/test/helpers';
import { Wrapper } from '.';

describe('components/container', () => {
  describe('<Wrapper />', () => {
    it('should render the wrapper styles correct', () => {
      const { container } = renderWithProviders(<Wrapper />);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
