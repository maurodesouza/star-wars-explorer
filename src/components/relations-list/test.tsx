import { screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test/helpers';

import { events } from 'app';

import { mock } from './mock';
import { RelationsList } from '.';
import { act } from '@testing-library/react-hooks';

jest.mock('components', () => ({
  __esModule: true,

  RelationsListFilter: function Mock() {
    return <div data-testid="relations list filter" />;
  },

  CardList: function Mock() {
    return <div data-testid="card list" />;
  },
}));

const dataMock = {
  ...mock,
};

describe('<RelationsList />', () => {
  it('should render the empty card styles correct', () => {
    const { container } = renderWithProviders(<RelationsList {...dataMock} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should send relations set options event on render', () => {
    const senOptionEvent = jest.spyOn(events.relations, 'set.options');

    renderWithProviders(<RelationsList {...dataMock} />);

    const entities = dataMock.relations.map(entity => entity.entity);

    expect(senOptionEvent).toBeCalledTimes(1);
    expect(senOptionEvent).toBeCalledWith(entities);
  });

  it('should filter list when receive relation filter event', () => {
    renderWithProviders(<RelationsList {...dataMock} />);

    let cards = screen.getAllByTestId('card list');

    const entityOne = dataMock.relations[0].entity;
    const entityTwo = dataMock.relations[1].entity;

    let labelOne = screen.queryByText(`${entityOne} (1)`);
    let labelTwo = screen.queryByText(`${entityTwo} (1)`);

    expect(cards.length).toBe(2);

    expect(labelOne).toBeInTheDocument();
    expect(labelTwo).toBeInTheDocument();

    act(() => {
      events.relations.filter(entityOne);
    });

    cards = screen.getAllByTestId('card list');

    labelOne = screen.queryByText(`${entityOne} (1)`);
    labelTwo = screen.queryByText(`${entityTwo} (1)`);

    expect(labelOne).toBeInTheDocument();
    expect(labelTwo).not.toBeInTheDocument();

    expect(cards.length).toBe(1);
  });

  it('should not render if no one realtions', () => {
    const { container } = renderWithProviders(<RelationsList />);

    expect(container.children.length).toBe(0);
  });

  it('should render corret count of finded cards case has empty card in relations', () => {
    const emptyCount = 2;

    const normalItem = dataMock.relations[0];
    const entityOne = normalItem.entity;

    const empties = Array(emptyCount)
      .fill('')
      .map((_, index) => ({
        id: index,
        entity: entityOne,
        empty: true,
      }));

    const relations = [normalItem, ...empties] as any[];

    renderWithProviders(<RelationsList relations={relations} />);

    const total = relations.length;
    const finded = total - emptyCount;

    const label = screen.queryByText(`${entityOne} (${finded}/${total})`);

    expect(label).toBeInTheDocument();
  });
});
