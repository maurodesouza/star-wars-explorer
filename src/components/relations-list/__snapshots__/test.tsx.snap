// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RelationsList /> should render the empty card styles correct 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 2.4rem;
  width: 100%;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 3.2rem;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1.2rem;
}

.c3 {
  font-size: 2.4rem;
  text-transform: capitalize;
}

<div
  class="c0"
>
  <div
    data-testid="relations list filter"
  />
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <h4
        class="c3"
      >
        characters (1)
      </h4>
      <div
        data-testid="card list"
      />
    </div>
    <div
      class="c2"
    >
      <h4
        class="c3"
      >
        films (1)
      </h4>
      <div
        data-testid="card list"
      />
    </div>
  </div>
</div>
`;
